envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/config/envoy/EnvoyExtension.kt

```diff
@@ -0,0 +1,46 @@
+package pl.allegro.tech.servicemesh.envoycontrol.config.envoy
+
+import org.junit.jupiter.api.extension.AfterAllCallback
+import org.junit.jupiter.api.extension.BeforeAllCallback
+import org.junit.jupiter.api.extension.ExtensionContext
+import org.testcontainers.containers.Network
+import pl.allegro.tech.servicemesh.envoycontrol.config.EnvoyControlExtension
+import pl.allegro.tech.servicemesh.envoycontrol.config.RandomConfigFile
+import pl.allegro.tech.servicemesh.envoycontrol.config.containers.EchoServiceExtension
+import pl.allegro.tech.servicemesh.envoycontrol.logger
+
+class EnvoyExtension(
+    envoyControl: EnvoyControlExtension,
+    val localService: EchoServiceExtension? = null
+) : BeforeAllCallback, AfterAllCallback {
+
+    companion object {
+        val logger by logger()
+    }
+
+    val container: EnvoyContainer = EnvoyContainer(
+            RandomConfigFile,
+            { localService?.container?.ipAddress() ?: "127.0.0.1" },
+            envoyControl.app.grpcPort
+    ).withNetwork(Network.SHARED)
+
+    val ingressOperations: IngressOperations = IngressOperations(container)
+    val egressOperations: EgressOperations = EgressOperations(container)
+
+    override fun beforeAll(context: ExtensionContext?) {
+        if (localService != null && !localService.started) {
+            localService.beforeAll(context)
```

Comment:
It won't be called twice, because it's guarded by a flag:
```
override fun beforeAll(context: ExtensionContext?) {
        if (started) {
            return
        }

        container.start()
        started = true
    }
```

It's about dependencies between extensions, some should be started before the others. I had 2 options for solving that:
* @Order annotations in every test class, on fields where you declare extensions.
* This option of making sure the extensions control that things start in the correct order.

I will hide this under a cleaner API in the future.